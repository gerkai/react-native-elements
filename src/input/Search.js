import PropTypes from 'prop-types';
import React, { Component } from 'react';
import {
  ActivityIndicator,
  View,
  StyleSheet,
  TextInput,
  Platform,
  Text as NativeText,
} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import colors from '../config/colors';
import normalize from '../helpers/normalizeText';

class Search extends Component {
  focus() {
    const ref = this.props.textInputRef;
    this.refs[ref].focus();
  }

  clearText() {
    const ref = this.props.textInputRef;
    this.refs[ref].clear();
  }

  render() {
    const {
      containerStyle,
      inputStyle,
      icon,
      noIcon,
      lightTheme,
      round,
      showLoadingIcon,
      loadingIcon,
      clearIcon,
      containerRef,
      textInputRef,
      selectionColor,
      underlineColorAndroid,
      ...attributes
    } = this.props;
    return (
      <View
        ref={containerRef}
        style={[
          styles.container,
          lightTheme && styles.containerLight,
          containerStyle && containerStyle,
        ]}
      >
        <TextInput
          ref={textInputRef}
          selectionColor={selectionColor || colors.grey3}
          underlineColorAndroid={
            underlineColorAndroid ? underlineColorAndroid : 'transparent'
          }
          style={[
            styles.input,
            lightTheme && styles.inputLight,
            noIcon && { paddingLeft: 9 },
            round && { borderRadius: Platform.OS === 'ios' ? 15 : 20 },
            inputStyle && inputStyle,
          ]}
          {...attributes}
        />
        {!noIcon &&
          <Icon
            size={16}
            style={[styles.icon, icon.style && icon.style]}
            name={icon.name || 'search'}
            color={icon.color || colors.grey3}
          />}
        {clearIcon &&
          <Icon
            size={16}
            style={[styles.clearIcon, clearIcon.style && clearIcon.style]}
            name={clearIcon.name || 'close'}
            onPress={this.clearText.bind(this)}
            color={clearIcon.color || colors.grey3}
          />}
        {showLoadingIcon &&
          <ActivityIndicator
            style={[styles.loadingIcon, loadingIcon.style && loadingIcon.style]}
            color={icon.color || colors.grey3}
          />}
      </View>
    );
  }
}

Search.propTypes = {
  icon: PropTypes.object,
  noIcon: PropTypes.bool,
  lightTheme: PropTypes.bool,
  containerStyle: View.propTypes.style,
  inputStyle: NativeText.propTypes.style,
  round: PropTypes.bool,
  showLoadingIcon: PropTypes.bool,
  loadingIcon: PropTypes.object,
  clearIcon: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
  textInputRef: PropTypes.string,
  containerRef: PropTypes.string,
  selectionColor: PropTypes.string,
  underlineColorAndroid: PropTypes.string,
};

Search.defaultProps = {
  placeholderTextColor: colors.grey3,
  lightTheme: false,
  noIcon: false,
  round: false,
  icon: {},
  showLoadingIcon: false,
  loadingIcon: {},
};

const styles = StyleSheet.create({
  container: {
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderBottomColor: '#000',
    borderTopColor: '#000',
    backgroundColor: colors.grey0,
  },
  containerLight: {
    backgroundColor: colors.grey5,
    borderTopColor: '#e1e1e1',
    borderBottomColor: '#e1e1e1',
  },
  icon: {
    backgroundColor: 'transparent',
    position: 'absolute',
    left: 16,
    top: 15.5,
    ...Platform.select({
      android: {
        top: 20,
      },
    }),
  },
  loadingIcon: {
    backgroundColor: 'transparent',
    position: 'absolute',
    right: 16,
    top: 13,
    ...Platform.select({
      android: {
        top: 17,
      },
    }),
  },
  input: {
    paddingLeft: 26,
    paddingRight: 19,
    margin: 8,
    borderRadius: 3,
    overflow: 'hidden',
    backgroundColor: colors.searchBg,
    fontSize: normalize(14),
    color: colors.grey3,
    height: 40,
    ...Platform.select({
      ios: {
        height: 30,
      },
      android: {
        borderWidth: 0,
      },
    }),
  },
  inputLight: {
    backgroundColor: colors.grey4,
  },
  clearIcon: {
    backgroundColor: 'transparent',
    position: 'absolute',
    right: 16,
    top: 15.5,
    ...Platform.select({
      android: {
        top: 17,
      },
    }),
  },
});

export default Search;
