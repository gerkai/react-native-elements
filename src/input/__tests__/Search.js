import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Search from '../Search';

describe('Search component', () => {
  it('should render without issues', () => {
    const component = shallow(<Search />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with icons', () => {
    const component = shallow(
      <Search
        clearIcon={{
          name: '3d-rotation',
          color: 'red',
        }}
        showLoadingIcon
        loadingIcon={{
          style: { flex: 1 },
        }}
        lightTheme
        containerStyle={{ height: 70 }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render without icon', () => {
    const component = shallow(
      <Search underlineColorAndroid="red" noIcon round />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });
});
