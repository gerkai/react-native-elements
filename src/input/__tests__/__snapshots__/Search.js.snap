// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component should render without issues 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <View
    style={
        Array [
            Object {
              "backgroundColor": "#393e42",
              "borderBottomColor": "#000",
              "borderBottomWidth": 1,
              "borderTopColor": "#000",
              "borderTopWidth": 1,
            },
            false,
            undefined,
          ]
    }
>
    <TextInput
        autoCapitalize={undefined}
        autoCorrect={undefined}
        autoFocus={undefined}
        blurOnSubmit={undefined}
        clear={undefined}
        clearButtonMode={undefined}
        clearTextOnFocus={undefined}
        dataDetectorTypes={undefined}
        defaultValue={undefined}
        editable={undefined}
        enablesReturnKeyAutomatically={undefined}
        inlineImageLeft={undefined}
        inlineImagePadding={undefined}
        isFocused={undefined}
        keyboardAppearance={undefined}
        keyboardType={undefined}
        maxLength={undefined}
        multiline={undefined}
        numberOfLines={undefined}
        onBlur={undefined}
        onChange={undefined}
        onChangeText={undefined}
        onContentSizeChange={undefined}
        onEndEditing={undefined}
        onFocus={undefined}
        onKeyPress={undefined}
        onLayout={undefined}
        onSelectionChange={undefined}
        onSubmitEditing={undefined}
        placeholder={undefined}
        placeholderTextColor="#86939e"
        returnKeyLabel={undefined}
        returnKeyType={undefined}
        secureTextEntry={undefined}
        selectTextOnFocus={undefined}
        selectionColor="#86939e"
        selectionState={undefined}
        style={
            Array [
                Object {
                  "backgroundColor": "#303337",
                  "borderRadius": 3,
                  "color": "#86939e",
                  "fontSize": 17.5,
                  "height": 30,
                  "margin": 8,
                  "overflow": "hidden",
                  "paddingLeft": 26,
                  "paddingRight": 19,
                },
                false,
                false,
                false,
                undefined,
              ]
        }
        underlineColorAndroid="transparent"
        value={undefined}
    />
    <Icon
        allowFontScaling={false}
        color="#86939e"
        name="search"
        size={16}
        style={
            Array [
                Object {
                  "backgroundColor": "transparent",
                  "left": 16,
                  "position": "absolute",
                  "top": 15.5,
                },
                undefined,
              ]
        }
    />
</View>,
  "nodes": Array [
    <View
      style={
            Array [
                  Object {
                    "backgroundColor": "#393e42",
                    "borderBottomColor": "#000",
                    "borderBottomWidth": 1,
                    "borderTopColor": "#000",
                    "borderTopWidth": 1,
                  },
                  false,
                  undefined,
                ]
      }
>
      <TextInput
            autoCapitalize={undefined}
            autoCorrect={undefined}
            autoFocus={undefined}
            blurOnSubmit={undefined}
            clear={undefined}
            clearButtonMode={undefined}
            clearTextOnFocus={undefined}
            dataDetectorTypes={undefined}
            defaultValue={undefined}
            editable={undefined}
            enablesReturnKeyAutomatically={undefined}
            inlineImageLeft={undefined}
            inlineImagePadding={undefined}
            isFocused={undefined}
            keyboardAppearance={undefined}
            keyboardType={undefined}
            maxLength={undefined}
            multiline={undefined}
            numberOfLines={undefined}
            onBlur={undefined}
            onChange={undefined}
            onChangeText={undefined}
            onContentSizeChange={undefined}
            onEndEditing={undefined}
            onFocus={undefined}
            onKeyPress={undefined}
            onLayout={undefined}
            onSelectionChange={undefined}
            onSubmitEditing={undefined}
            placeholder={undefined}
            placeholderTextColor="#86939e"
            returnKeyLabel={undefined}
            returnKeyType={undefined}
            secureTextEntry={undefined}
            selectTextOnFocus={undefined}
            selectionColor="#86939e"
            selectionState={undefined}
            style={
                  Array [
                        Object {
                          "backgroundColor": "#303337",
                          "borderRadius": 3,
                          "color": "#86939e",
                          "fontSize": 17.5,
                          "height": 30,
                          "margin": 8,
                          "overflow": "hidden",
                          "paddingLeft": 26,
                          "paddingRight": 19,
                        },
                        false,
                        false,
                        false,
                        undefined,
                      ]
            }
            underlineColorAndroid="transparent"
            value={undefined}
      />
      <Icon
            allowFontScaling={false}
            color="#86939e"
            name="search"
            size={16}
            style={
                  Array [
                        Object {
                          "backgroundColor": "transparent",
                          "left": 16,
                          "position": "absolute",
                          "top": 15.5,
                        },
                        undefined,
                      ]
            }
      />
</View>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Search
        icon={Object {}}
        lightTheme={false}
        loadingIcon={Object {}}
        noIcon={false}
        placeholderTextColor="#86939e"
        round={false}
        showLoadingIcon={false}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Search {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "icon": Object {},
          "lightTheme": false,
          "loadingIcon": Object {},
          "noIcon": false,
          "placeholderTextColor": "#86939e",
          "round": false,
          "showLoadingIcon": false,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <View
          style={
                    Array [
                              Object {
                                "backgroundColor": "#393e42",
                                "borderBottomColor": "#000",
                                "borderBottomWidth": 1,
                                "borderTopColor": "#000",
                                "borderTopWidth": 1,
                              },
                              false,
                              undefined,
                            ]
          }
>
          <TextInput
                    autoCapitalize={undefined}
                    autoCorrect={undefined}
                    autoFocus={undefined}
                    blurOnSubmit={undefined}
                    clear={undefined}
                    clearButtonMode={undefined}
                    clearTextOnFocus={undefined}
                    dataDetectorTypes={undefined}
                    defaultValue={undefined}
                    editable={undefined}
                    enablesReturnKeyAutomatically={undefined}
                    inlineImageLeft={undefined}
                    inlineImagePadding={undefined}
                    isFocused={undefined}
                    keyboardAppearance={undefined}
                    keyboardType={undefined}
                    maxLength={undefined}
                    multiline={undefined}
                    numberOfLines={undefined}
                    onBlur={undefined}
                    onChange={undefined}
                    onChangeText={undefined}
                    onContentSizeChange={undefined}
                    onEndEditing={undefined}
                    onFocus={undefined}
                    onKeyPress={undefined}
                    onLayout={undefined}
                    onSelectionChange={undefined}
                    onSubmitEditing={undefined}
                    placeholder={undefined}
                    placeholderTextColor="#86939e"
                    returnKeyLabel={undefined}
                    returnKeyType={undefined}
                    secureTextEntry={undefined}
                    selectTextOnFocus={undefined}
                    selectionColor="#86939e"
                    selectionState={undefined}
                    style={
                              Array [
                                        Object {
                                          "backgroundColor": "#303337",
                                          "borderRadius": 3,
                                          "color": "#86939e",
                                          "fontSize": 17.5,
                                          "height": 30,
                                          "margin": 8,
                                          "overflow": "hidden",
                                          "paddingLeft": 26,
                                          "paddingRight": 19,
                                        },
                                        false,
                                        false,
                                        false,
                                        undefined,
                                      ]
                    }
                    underlineColorAndroid="transparent"
                    value={undefined}
          />
          <Icon
                    allowFontScaling={false}
                    color="#86939e"
                    name="search"
                    size={16}
                    style={
                              Array [
                                        Object {
                                          "backgroundColor": "transparent",
                                          "left": 16,
                                          "position": "absolute",
                                          "top": 15.5,
                                        },
                                        undefined,
                                      ]
                    }
          />
</View>,
        "_debugID": 2,
        "_renderedOutput": <View
          style={
                    Array [
                              Object {
                                "backgroundColor": "#393e42",
                                "borderBottomColor": "#000",
                                "borderBottomWidth": 1,
                                "borderTopColor": "#000",
                                "borderTopWidth": 1,
                              },
                              false,
                              undefined,
                            ]
          }
>
          <TextInput
                    autoCapitalize={undefined}
                    autoCorrect={undefined}
                    autoFocus={undefined}
                    blurOnSubmit={undefined}
                    clear={undefined}
                    clearButtonMode={undefined}
                    clearTextOnFocus={undefined}
                    dataDetectorTypes={undefined}
                    defaultValue={undefined}
                    editable={undefined}
                    enablesReturnKeyAutomatically={undefined}
                    inlineImageLeft={undefined}
                    inlineImagePadding={undefined}
                    isFocused={undefined}
                    keyboardAppearance={undefined}
                    keyboardType={undefined}
                    maxLength={undefined}
                    multiline={undefined}
                    numberOfLines={undefined}
                    onBlur={undefined}
                    onChange={undefined}
                    onChangeText={undefined}
                    onContentSizeChange={undefined}
                    onEndEditing={undefined}
                    onFocus={undefined}
                    onKeyPress={undefined}
                    onLayout={undefined}
                    onSelectionChange={undefined}
                    onSubmitEditing={undefined}
                    placeholder={undefined}
                    placeholderTextColor="#86939e"
                    returnKeyLabel={undefined}
                    returnKeyType={undefined}
                    secureTextEntry={undefined}
                    selectTextOnFocus={undefined}
                    selectionColor="#86939e"
                    selectionState={undefined}
                    style={
                              Array [
                                        Object {
                                          "backgroundColor": "#303337",
                                          "borderRadius": 3,
                                          "color": "#86939e",
                                          "fontSize": 17.5,
                                          "height": 30,
                                          "margin": 8,
                                          "overflow": "hidden",
                                          "paddingLeft": 26,
                                          "paddingRight": 19,
                                        },
                                        false,
                                        false,
                                        false,
                                        undefined,
                                      ]
                    }
                    underlineColorAndroid="transparent"
                    value={undefined}
          />
          <Icon
                    allowFontScaling={false}
                    color="#86939e"
                    name="search"
                    size={16}
                    style={
                              Array [
                                        Object {
                                          "backgroundColor": "transparent",
                                          "left": 16,
                                          "position": "absolute",
                                          "top": 15.5,
                                        },
                                        undefined,
                                      ]
                    }
          />
</View>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Search
    icon={Object {}}
    lightTheme={false}
    loadingIcon={Object {}}
    noIcon={false}
    placeholderTextColor="#86939e"
    round={false}
    showLoadingIcon={false}
/>,
}
`;
